session/timerq.go

```diff
@@ -0,0 +1,153 @@
+// timerq.go - Time delayed queue
+// Copyright (C) 2018  Masala, David Stainton.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package session
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"container/heap"
+	"github.com/katzenpost/core/queue"
+	"github.com/katzenpost/core/worker"
+)
+
+type nqueue interface {
+	Push(*Message) error
+}
+
+// TimerQ is a queue that delays messages before forwarding to another queue
+type TimerQ struct {
+	sync.Mutex
+	sync.Cond
+	worker.Worker
+
+	priq  *queue.PriorityQueue
+	nextQ nqueue
+
+	timer  *time.Timer
+	wakech chan struct{}
+}
+
+// NewTimerQ intantiates a new TimerQ and starts the worker routine
+func NewTimerQ(q nqueue) *TimerQ {
+	a := &TimerQ{
+		nextQ: q,
+		timer: time.NewTimer(0),
+		priq:  queue.New(),
+	}
+	a.L = new(sync.Mutex)
+	a.Go(a.worker)
+	return a
+}
+
+// Push adds a message to the TimerQ
+func (a *TimerQ) Push(m *Message) {
+	a.Lock()
+	a.priq.Enqueue(m.expiry(), m)
+	a.Unlock()
+	a.Signal()
+}
+
+// Remove removes a Message from the TimerQ
+func (a *TimerQ) Remove(m *Message) error {
+	a.Lock()
+	defer a.Unlock()
+	if mo := a.priq.Peek(); mo != nil {
+		if mo.Value.(*Message) == m {
```

Comment:
if the equality is false then what? the queue item will not get removed. this seems wrong.